Instances with type-providers are not assignable to `FastifyInstance`

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the bug has not already been reported


### Fastify version

4.7.0

### Plugin version

_No response_

### Node.js version

18.1.0

### Operating system

Linux

### Operating system version (i.e. 20.04, 11.3, 10)

Ubuntu 20.04.5 LTS

### Description

Fastify instances with an enabled type-provider are not subtypes of the base `FastifyInstance` interface.

## Example

A fastify server with type-providers enabled will not be usable as input for the `@fastify/aws-lambda` plugin, which expects a `FastifyInstance` parameter.

```ts
export default function awsLambdaFastify<TEvent, TResult = LambdaResponse>(
  app: FastifyInstance,
  options?: LambdaFastifyOptions
): PromiseHandler<TEvent, TResult>;
```

## Cause
A more detailed explanation can be found in the [reproduction repo](https://github.com/driimus/fastify-instance-assignability#cause). 

In short, the assignability issue stems from certain methods of `FastifyInstance` being contravariant on `TypeProvider`. Because of that, the compiler has to flip the assignability check's direction, which boils down to the following outcome:

```diff
Types of property 'handler' are incompatible.
  Type 'RouteHandlerMethod<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, ... 4 more ..., FastifyBaseLogger>' is not assignable to type 'RouteHandlerMethod<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, any, any, any, FastifyTypeProviderDefault, FastifyBaseLogger>'.
    Types of parameters 'request' and 'request' are incompatible.
      Type 'FastifyRequest<any, RawServerDefault, IncomingMessage, any, FastifyTypeProviderDefault, any, FastifyBaseLogger, ResolveFastifyRequestType<...>>' is not assignable to type 'FastifyRequest<any, Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, any, TypeBoxTypeProvider, any, FastifyBaseLogger, ResolveFastifyRequestType<...>>'.
        Type 'ResolveFastifyRequestType<FastifyTypeProviderDefault, any, any>' is not assignable to type 'ResolveFastifyRequestType<TypeBoxTypeProvider, any, any>'.
+          Type 'FastifyTypeProviderDefault' is not assignable to type 'TypeBoxTypeProvider'.
```


### Steps to Reproduce

1. Check out the [Minimal reproduction repo](https://github.com/driimus/fastify-instance-assignability), which uses the `"@fastify/aws-lambda": "^3.1.3"` plugin.
1. Install dependencies
1. Run `pnpm tsc --noEmit` to check for type errors
1. Notice the assignability error 

From the CI run: https://github.com/driimus/fastify-instance-assignability/actions/runs/3259362782/jobs/5352129645#step:6:11
```
Error: src/index.ts([27](https://github.com/driimus/fastify-instance-assignability/actions/runs/3259362782/jobs/5352129645#step:6:28),41): error TS2769: No overload matches this call.
  Overload 1 of 2, '(app: FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>, options?: LambdaFastifyOptions | undefined): PromiseHandler<...>', gave the following error.
    Argument of type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider>' is not assignable to parameter of type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>'.
      The types returned by 'after()' are incompatible between these types.
        Type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider> & PromiseLike<...>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault> & PromiseLike<...>'.
          Type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider> & PromiseLike<...>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>'.
            Types of property 'register' are incompatible.
              Type 'FastifyRegister<FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider> & PromiseLike<...>, RawServerDefault, FastifyTypeProvider>' is not assignable to type 'FastifyRegister<FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault> & PromiseLike<...>, RawServerDefault, FastifyTypeProvider>'.
                Type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider> & PromiseLike<...>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault> & PromiseLike<...>'.
                  Type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider> & PromiseLike<...>' is not assignable to type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>'.
                    The types of 'addSchema(...).addHook' are incompatible between these types.
                      Type '{ <RouteGeneric extends import("/home/runner/work/fastify-instance-assignability/fastify-instance-assignability/node_modules/.pnpm/fastify@4.8.1/node_modules/fastify/types/route").RouteGenericInterface = import("/home/runner/work/fastify-instance-assignability/fastify-instance-assignability/node_modules/.pnpm/fastif...' is not assignable to type '{ <RouteGeneric extends import("/home/runner/work/fastify-instance-assignability/fastify-instance-assignability/node_modules/.pnpm/fastify@4.8.1/node_modules/fastify/types/route").RouteGenericInterface = import("/home/runner/work/fastify-instance-assignability/fastify-instance-assignability/node_modules/.pnpm/fastif...'. Two different types with this name exist, but they are unrelated.
                        Types of parameters 'hook' and 'hook' are incompatible.
                          Types of parameters 'opts' and 'opts' are incompatible.
                            Type 'RouteOptions<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, ... 4 more ..., FastifyBaseLogger> & { ...; }' is not assignable to type 'RouteOptions<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, any, any, any, FastifyTypeProviderDefault, FastifyBaseLogger> & { ...; }'.
                              Type 'RouteOptions<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, ... 4 more ..., FastifyBaseLogger> & { ...; }' is not assignable to type 'RouteOptions<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, any, any, any, FastifyTypeProviderDefault, FastifyBaseLogger>'.
                                Types of property 'handler' are incompatible.
                                  Type 'RouteHandlerMethod<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, ... 4 more ..., FastifyBaseLogger>' is not assignable to type 'RouteHandlerMethod<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, any, any, any, FastifyTypeProviderDefault, FastifyBaseLogger>'.
                                    Types of parameters 'request' and 'request' are incompatible.
                                      Type 'FastifyRequest<any, RawServerDefault, IncomingMessage, any, FastifyTypeProviderDefault, any, FastifyBaseLogger, ResolveFastifyRequestType<...>>' is not assignable to type 'FastifyRequest<any, Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, any, TypeBoxTypeProvider, any, FastifyBaseLogger, ResolveFastifyRequestType<...>>'.
                                        Type 'ResolveFastifyRequestType<FastifyTypeProviderDefault, any, any>' is not assignable to type 'ResolveFastifyRequestType<TypeBoxTypeProvider, any, any>'.
                                          Type 'FastifyTypeProviderDefault' is not assignable to type 'TypeBoxTypeProvider'.
  Overload 2 of 2, '(app: FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>, options?: LambdaFastifyOptions | undefined): CallbackHandler<...>', gave the following error.
    Argument of type 'FastifyInstance<Server<typeof IncomingMessage, typeof ServerResponse>, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, TypeBoxTypeProvider>' is not assignable to parameter of type 'FastifyInstance<RawServerDefault, IncomingMessage, ServerResponse<IncomingMessage>, FastifyBaseLogger, FastifyTypeProviderDefault>'.
```

### Expected Behavior

I would expect specialised versions of the `FastifyInstance` interface, such as the ones using a custom type provider, to be assignable to `FastifyInstance`.

I am aware of some available workarounds:
1. Extracting the type-provider enabled instance into a separate reference (similar to the [scoped type-provider](https://www.fastify.io/docs/latest/Reference/Type-Providers/#scoped-type-provider) docs suggestion).
1. Extending affected function signatures to include generic parameters which are propagated to their input argument of type `FastifyInstance`.

However, both of them seem inferior to ensuring that subtypes of `FastifyInstance` are assignable to the base interface. 
I haven't look at the way type providers are being used in details, but it seems possible: https://github.com/fastify/fastify/compare/77b7582...driimus:fastify:test/type-provider-contravariance

--- a/docs/Guides/Write-Type-Provider.md
+++ b/docs/Guides/Write-Type-Provider.md
@@ -1,0 +1,32 @@
+<h1 align="center">Fastify</h1>
+
+## How to write your own type provider
+
+Things to keep in mind when implementing a custom [type provider](../Reference/Type-Providers.md):
+
+### Type Contravariance
+
+Whereas exhaustive type narrowing checks normally rely on `never` to represent
+an unreachable state, reduction in type provider interfaces should only be done
+up to `unknown`.
+
+The reasoning is that certain methods of `FastifyInstance` are 
+contravariant on `TypeProvider`, which can lead to TypeScript surfacing 
+assignability issues unless the custom type provider interface is 
+substitutible with `FastifyTypeProviderDefault`.
+
+For example, `FastifyTypeProviderDefault` will not be assignable to the following:
+```ts
+export interface NotSubstitutibleTypeProvider extends FastifyTypeProvider {
+   // bad, nothing is assignable to `never` (except for itself)
+  output: this['input'] extends /** custom check here**/ ? /** narrowed type here **/ : never;
+}
+```
+
+Unless changed to:
+```ts
+export interface SubstitutibleTypeProvider extends FastifyTypeProvider {
+  // good, anything can be assigned to `unknown`
+  output: this['input'] extends /** custom check here**/ ? /** narrowed type here **/ : unknown; 
+}
+```

--- a/test/types/type-provider.test-d.ts
+++ b/test/types/type-provider.test-d.ts
@@ -68,7 +68,7 @@ expectAssignable(server.withTypeProvider<OverriddenProvider>().get<{ Body: 'over
 // TypeBox
 // -------------------------------------------------------------------
 
-interface TypeBoxProvider extends FastifyTypeProvider { output: this['input'] extends TSchema ? Static<this['input']> : never }
+interface TypeBoxProvider extends FastifyTypeProvider { output: this['input'] extends TSchema ? Static<this['input']> : unknown }
 
 expectAssignable(server.withTypeProvider<TypeBoxProvider>().get(
   '/',
@@ -94,7 +94,7 @@ expectAssignable<FastifyInstance>(server.withTypeProvider<TypeBoxProvider>())
 // JsonSchemaToTs
 // -------------------------------------------------------------------
 
-interface JsonSchemaToTsProvider extends FastifyTypeProvider { output: this['input'] extends JSONSchema ? FromSchema<this['input']> : never }
+interface JsonSchemaToTsProvider extends FastifyTypeProvider { output: this['input'] extends JSONSchema ? FromSchema<this['input']> : unknown }
 
 expectAssignable(server.withTypeProvider<JsonSchemaToTsProvider>().get(
   '/',

--- a/test/types/type-provider.test-d.ts
+++ b/test/types/type-provider.test-d.ts
@@ -88,6 +88,8 @@ expectAssignable(server.withTypeProvider<TypeBoxProvider>().get(
   }
 ))
 
+expectAssignable<FastifyInstance>(server.withTypeProvider<TypeBoxProvider>())
+
 // -------------------------------------------------------------------
 // JsonSchemaToTs
 // -------------------------------------------------------------------
@@ -114,6 +116,8 @@ expectAssignable(server.withTypeProvider<JsonSchemaToTsProvider>().get(
     expectType<boolean | undefined>(req.body.z)
   }
 ))
+
+expectAssignable<FastifyInstance>(server.withTypeProvider<JsonSchemaToTsProvider>())
 
 // -------------------------------------------------------------------
 // Instance Type Remappable

